src/librustc_mir/transform/const_prop.rs

```diff
@@ -620,14 +605,22 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             // if an overflow would occur.
             Rvalue::UnaryOp(UnOp::Neg, arg) if !overflow_check => {
                 trace!("checking UnaryOp(op = Neg, arg = {:?})", arg);
-                self.check_unary_op(arg, source_info)?;
+                // If `overflow_check` is set, running const-prop on the `Assert` terminators
+                // will already generate the appropriate messages.
+                if !overflow_check {
```

Comment:
*oops* this is redundant, I had missed the `if !overflow_check` match guard. That also explains why `Neg` overflow reporting behaved as expected (with no duplicates) before.